amazon/ion/ioncmodule.c

```diff
@@ -900,122 +902,89 @@ static PyObject* ionc_get_timestamp_precision(int precision) {
-                if (decContextTestStatus(&dec_context, DEC_Inexact)) {
-                    // This means the fractional component is not [0, 1) or has more than microsecond precision.
-                    decContextClearStatus(&dec_context, DEC_Inexact);
-                }
-
-                char dec_num[DECQUAD_String];
-                decQuad d;
-                decQuadFromInt32(&d, dec);
-                decQuadScaleB(&d, &d, decQuadFromInt32(&tmp, -fractional_precision), &dec_context);
-                decQuadToString(&d, dec_num);
-
-                PyObject* py_dec_str = PyUnicode_FromString(dec_num);
-                PyObject* py_fractional_seconds = PyObject_CallFunctionObjArgs(_decimal_constructor, py_dec_str, NULL);
-                PyDict_SetItemString(timestamp_args, "fractional_seconds", py_fractional_seconds);
-                Py_DECREF(py_fractional_seconds);
-                Py_DECREF(py_dec_str);
-            } else {
-                decQuadScaleB(&fraction, &fraction, decQuadFromInt32(&tmp, MICROSECOND_DIGITS), &dec_context);
-                int32_t microsecond = decQuadToInt32Exact(&fraction, &dec_context, DEC_ROUND_DOWN);
+            decQuadScaleB(&fraction, &fraction, decQuadFromInt32(&tmp, fractional_precision), &dec_context);
+            int dec = decQuadToInt32Exact(&fraction, &dec_context, DEC_ROUND_DOWN);
+            if (fractional_precision > MAX_TIMESTAMP_PRECISION) fractional_precision = MAX_TIMESTAMP_PRECISION;
+            if (decContextTestStatus(&dec_context, DEC_Inexact)) {
+                // This means the fractional component is not [0, 1) or has more than microsecond precision.
+                decContextClearStatus(&dec_context, DEC_Inexact);
+            }
 
-                if (decContextTestStatus(&dec_context, DEC_Inexact)) {
-                    // This means the fractional component is not [0, 1) or has more than microsecond precision.
-                    decContextClearStatus(&dec_context, DEC_Inexact);
-                }
+            char dec_num[DECQUAD_String];
+            decQuad d;
+            decQuadFromInt32(&d, dec);
+            decQuadScaleB(&d, &d, decQuadFromInt32(&tmp, -fractional_precision), &dec_context);
+            decQuadToString(&d, dec_num);
 
-                PyObject* py_microsecond = PyLong_FromLong(microsecond);
-                PyObject* py_fractional_precision = PyLong_FromLong(fractional_precision);
-                PyDict_SetItemString(timestamp_args, "microsecond", py_microsecond);
-                PyDict_SetItemString(timestamp_args, "fractional_precision", py_fractional_precision);
-                Py_DECREF(py_microsecond);
-                Py_DECREF(py_fractional_precision);
-            }
+            if (fractional_precision > MICROSECOND_DIGITS) fractional_precision = MICROSECOND_DIGITS;
+            py_fractional_seconds = PyObject_CallFunction(_decimal_constructor, "s", dec_num, NULL);
```

Comment:
Are you asserting that using a string as the intermediary between decQuad and Decimal is faster than construction using coefficient/scale, or is it just less code? Either way, it's interesting.

Reply:
I was just looking at how to simplify. To make sure I understand your question though: we were already doing decQuad => string => Decimal for the arbitrary precision path, I just collapsed the "micros or less precise" path to that.

I did that because the `Timestamp.__new__` _must_ derive the microseconds if they are not provided (due to the python requirement that the `datetime` fields be set in `__new__` and not `__init__`). So for the "micros or less precise" path I would need to do the `fractional_seconds` and `py_fractional_precision` here anyway (deriving them in python is slow so I kept that out of `__new__`).

So, given that I had to get the `fractional_seconds` and `py_fractional_seconds` for either path here, _and_ had to derive `microsecond` in `__new__`, this seemed like the right mix of simplicity to performance.

Reply:
Makes sense, thanks.